vsintegration/src/FSharp.ProjectSystem.FSharp/ProjectFlagsService.fs

```diff
@@ -0,0 +1,128 @@
+    
+    /// Moves a project to the front of the queue if it is already in the queue.
+    member __.PrioritiseProject(project : ProjectNode) =
+        lock queueLock <| fun () ->
+            match dirtyProjects.TryGetValue project with
+            | true, true ->
+                projectAdded.Trigger project
+            | _ -> ()
+    
+    member __.Clear() =
+        lock queueLock <| fun () ->
+            dirtyProjects.Clear()
+    
+    member this.StartProcessing() =
+        let processProject (project : ProjectNode) = async {
+            let completeEvent = Event<unit>()
+            
+            // Keep trying to grab the exclusive lock for a design time build
+            while accessor.BeginDesignTimeBuild() |> ErrorHandler.Failed do
+                do! Async.Sleep 2000
+
+            // Notify the project to not invoke compilation when a call comes from Fsc. We're
+            // only interested in the source files and flags that are passed from MSBuild.
+            let buildResource = project.AcquireExclusiveBuildResource(actuallyBuild = false)
+
+            // Build completion callback - called from the UI thread
+            let onComplete (_result : MSBuildResult) (_instance : ProjectInstance) =
+                buildResource.Dispose()
+                project.NotifySourcesAndFlags()
+                accessor.EndDesignTimeBuild() |> ignore
+                completeEvent.Trigger()
+
+            // Notes regarding the _ResolveReferenceDependencies property
+            // TODO: check whether this can be achieved in a less painful way 
+            // (names that start with '_' are treated as implementation details and not recommended
+            // for usage).
+            // Setting this property enforces MSBuild to resolve second order dependencies, so if
+            // desktop applications references .NET Core based portable assemblies then MSBuild
+            // will detect that one of references requires System.Runtime and expand list of
+            // references with framework facades for portables.
+            // Usually facades are located at: %programfiles(x86)%\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\Facades\
+            // If the property is not set then MSBuild will resolve only primary dependencies, and
+            // the compiler will be very unhappy as all fundamental types should be taken from
+            // System.Runtime that is not supplied.
+            
+            let projectInstance = ref null
+            UIThread.Run(fun () ->
+                project.DoMSBuildSubmission(BuildKind.ASYNC, "Compile", projectInstance, MSBuildCoda onComplete, [KeyValuePair("_ResolveReferenceDependencies", "true")]) |> ignore
+                )
+            
+            // Wait until the build is complete before we continue servicing the queue
+            do! Async.AwaitEvent completeEvent.Publish
+        }
+
+        async {
+            while true do
```

Comment:
This basically looks very clear.  But I'd somehow like to see some safeguards where we try/catch any exceptions in chunks of this code.  Infinite loops starting mailbox processors make me nervous if exceptions might happen.  

And are we sure the solution events get raised reliably?  

Basically I'm wondering "what can go catastrophically wrong that can cause this service to malfunction in really bad ways"